File: 21-Greedy-Algorithms/cpp/candy.cpp
/*
 * Problem: Candy
 * Source:  LeetCode 135 - https://leetcode.com/problems/candy/
 * 
 * Approach: Two-pass greedy. Left-to-right: if ratings[i] > ratings[i-1], 
 *           give more candy. Right-to-left: same for right neighbor.
 *           Take max of both passes.
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 * 
 * Tags: greedy, arrays
 */

#include <bits/stdc++.h>
using namespace std;

int candy(vector<int>& ratings) {
    int n = ratings.size();
    vector<int> candies(n, 1);
    
    // Left to right pass
    for (int i = 1; i < n; i++) {
        if (ratings[i] > ratings[i - 1]) {
            candies[i] = candies[i - 1] + 1;
        }
    }
    
    // Right to left pass
    for (int i = n - 2; i >= 0; i--) {
        if (ratings[i] > ratings[i + 1]) {
            candies[i] = max(candies[i], candies[i + 1] + 1);
        }
    }
    
    return accumulate(candies.begin(), candies.end(), 0);
}

int main() {
    vector<int> ratings = {1, 0, 2};
    cout << "Min candies: " << candy(ratings) << endl; // Output: 5
    
    vector<int> ratings2 = {1, 2, 2};
    cout << "Min candies: " << candy(ratings2) << endl; // Output: 4
    
    return 0;
}

File: 21-Greedy-Algorithms/java/Candy.java
/*
 * Problem: Candy
 * Source:  LeetCode 135 - https://leetcode.com/problems/candy/
 * 
 * Approach: Two-pass greedy — left-to-right then right-to-left.
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 * 
 * Tags: greedy, arrays
 */

import java.util.*;

public class Candy {
    
    public int candy(int[] ratings) {
        int n = ratings.length;
        int[] candies = new int[n];
        Arrays.fill(candies, 1);
        
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) {
                candies[i] = candies[i - 1] + 1;
            }
        }
        
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                candies[i] = Math.max(candies[i], candies[i + 1] + 1);
            }
        }
        
        int total = 0;
        for (int c : candies) total += c;
        return total;
    }
    
    public static void main(String[] args) {
        Candy sol = new Candy();
        System.out.println("Min candies: " + sol.candy(new int[]{1, 0, 2})); // 5
        System.out.println("Min candies: " + sol.candy(new int[]{1, 2, 2})); // 4
    }
}
